File: vercel-ai/tour-of-agents/src/workflow-evaluator-optimizer.ts
import * as restate from "@restatedev/restate-sdk";
import { openai } from "@ai-sdk/openai";
import { generateText, wrapLanguageModel } from "ai";
import { durableCalls } from "@restatedev/vercel-ai-middleware";
import {CodeGenRequestSchema, CodeGenRequest} from "./utils/types";
const schema = restate.serde.schema;

// <start_here>
const generate = async (ctx: restate.Context, req: CodeGenRequest) => {
  const model = wrapLanguageModel({
    model: openai("gpt-4o"),
    middleware: durableCalls(ctx, { maxRetryAttempts: 3 }),
  });

  let feedback = "";
  const maxIterations = 3;

  for (let i = 0; i < maxIterations; i++) {
    // Step 1: Generate code
    const { text: code } = await generateText({
      model,
      system: "You are a code generator. Write clean, correct code.",
      prompt: feedback
        ? `Task: ${req.task}\n\nPrevious attempt was rejected:\n${feedback}\n\nPlease fix the issues.`
        : `Task: ${req.task}`,
    });

    // Step 2: Evaluate the code
    const { text: evaluation } = await generateText({
      model,
      system: `You are a code reviewer. Evaluate the code for correctness,
            readability, and edge cases. Respond with PASS if acceptable,
            or FAIL: <feedback> with specific issues to fix.`,
      prompt: `Task: ${req.task}\n\nCode:\n${code}`,
    });

    if (evaluation.startsWith("PASS")) {
      return { code, iterations: i + 1 };
    }

    feedback = evaluation;
  }

  return { code: "Max iterations reached", iterations: maxIterations };
};

const agent = restate.service({
  name: "CodeGenerator",
  handlers: {
    generate: restate.createServiceHandler(
      { input: schema(CodeGenRequestSchema) },
      generate,
    ),
  },
});
// <end_here>

restate.serve({ services: [agent] });

File: langchain-python/tour-of-agents/app/workflow_evaluator_optimizer.py
"""Generator/evaluator loop: a writer agent and a reviewer agent iterate
until the reviewer says PASS or we hit the iteration cap."""

import restate
from langchain.agents import create_agent
from langchain.chat_models import init_chat_model

from restate.ext.langchain import RestateMiddleware

from utils.models import CodeRequest


# <start_here>
generator = create_agent(
    model=init_chat_model("openai:gpt-4o-mini"),
    tools=[],
    system_prompt="You are a code generator. Write clean, correct code.",
    middleware=[RestateMiddleware()],
)

evaluator = create_agent(
    model=init_chat_model("openai:gpt-4o-mini"),
    tools=[],
    system_prompt=(
        "You are a code reviewer. Evaluate the code for correctness, "
        "readability, and edge cases. Respond with PASS if acceptable, or "
        "FAIL: <feedback> with specific issues to fix."
    ),
    middleware=[RestateMiddleware()],
)

code_service = restate.Service("CodeGenerator")


@code_service.handler()
async def generate(_ctx: restate.Context, req: CodeRequest) -> dict:
    feedback = ""
    max_iterations = 3

    for i in range(max_iterations):
        # Step 1: Generate code
        prompt = (
            f"Task: {req.task}\n\nPrevious attempt was rejected:\n{feedback}\n\nPlease fix the issues."
            if feedback
            else f"Task: {req.task}"
        )
        gen_result = await generator.ainvoke({"messages": [{"role": "user", "content": prompt}]})
        code = gen_result["messages"][-1].content

        # Step 2: Evaluate the code
        eval_result = await evaluator.ainvoke(
            {"messages": [{"role": "user", "content": f"Task: {req.task}\n\nCode:\n{code}"}]}
        )
        evaluation = eval_result["messages"][-1].content

        if evaluation.startswith("PASS"):
            return {"code": code, "iterations": i + 1}

        feedback = evaluation

    return {"code": "Max iterations reached", "iterations": max_iterations}


# <end_here>


if __name__ == "__main__":
    import asyncio

    import hypercorn
    import hypercorn.asyncio

    app = restate.app(services=[code_service])
    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))
